### ISSUE TITLE
provide a way to not download `tests/`, `examples` and other such common directories

### ISSUE THREAD

**Issue by poliorcetics** [rust-lang/cargo#13491]
### Problem

By default `cargo` includes everything from the root of the package (or almost, see https://doc.rust-lang.org/cargo/reference/manifest.html?highlight=files#the-exclude-and-include-fields)

This is very nice for archiving purposes by crates.io, it allows example scraping from docs.rs and probably other purposes I haven't encountered, but it has an issue: it makes lots of CI downloads heavier than they need to be.

Looking through the dependencies we use at $work I see around a 100MiB of unused files just counting `tests/`, but there are also `.github`, `benches/`, `examples/` and probably more I missed, and we have only ~600 deps out of the 138k crates on crates.io (without even counting all the available versions).

We have a cache to avoid spamming crates.io whenever possible so it's not like we request an extra 100MiB on each CI but still, I don't like putting pressure on it if we can avoid it.

As linked above, https://doc.rust-lang.org/cargo/reference/manifest.html?highlight=files#the-exclude-and-include-fields is intended to help with that but it's not automatic: people need to consider and maintain it for each crate (and from the start, else older versions will still have all the unused files)

### Proposed Solution

We may need a way to say to cargo `only fetch the absolute necessary to build the crate`. Two issues with that:

- What is the "absolute necessary": probably at least `src/, Cargo.toml, license files, build.rs` ? Some `-sys` crates may need more
- Changing that silently in a backward compatible way is probably hard ? I think it would break the hashes so it could not become the default automatically at the very least.

It could take the form of `cargo fetch --no-extras` (to be bikeshedded) ?

### Notes

Note: I'm excluding binary crates here, but they could easily be included too, if only so `cargo install` downloads less if possible, but I don't think binary crates are downloaded as much as library ones.

I also don't know if the benefits would be worth it, since it could mean one of:

- Recomputing the set of files on each download
- Storing two archives of the same crate&version

I have no data saying the bandwidth gains would be enough to offset either the compute or space gains, if anyone knows (the infra team ?) I would be happy to be proven wrong!

**Comment by epage** [comment]
For `cargo vendor`, we also have #13474.

Breaking down the extraneous files you mentioned:
- `.github` there is no reason to include and that is on the package for including it.
- For `examples/`, that can tie into the documentation and can be important for including.
- For `tests/` and `benches/`, that is important for crater and people wanting to do similar for crater.

The challenge is knowing what is needed for build or not among the files.  A heuristic can work but that can only go so far (much like we provide `package.include` and `package.exclude` in case our existing heuristic isn't enough.

There is also the question of "what is essential".  For example, explicitly referenced targets are sometimes required just to parse manifests today (see #13456).

There would need to be figured out to get this to work.

**Comment by weihanglo** [comment]
I would like to call out that [the `#[path]` attribute](https://doc.rust-lang.org/reference/items/modules.html#the-path-attribute), while not widely used, would cause problems if Cargo excludes files without recognizing where they point to. Besides, changing or disable the heuristics might affect other parts in Carog, like <https://github.com/rust-lang/cargo/issues/4587>, though this doesn't mean that Cargo cannot change the current rule.

**Comment by epage** [comment]
I don't think we re-check hashes but i would not rely on that never changing.

**Comment by poliorcetics** [comment]
> i would not rely on that never changing.

It should be a line or two of bash, it can be deleted if an update breaks it :)

**Comment by poliorcetics** [comment]
Update: cargo is too well-made and check hashes even in vendored dependencies so my two lines of bash won't work 😞

**Comment by poliorcetics** [comment]
A simpler way to do this then would be to allow not checking hashes when compiling ?

**Comment by newpavlov** [comment]
Relevant [IRLO discussion](https://internals.rust-lang.org/t/24188/12).

>.github there is no reason to include and that is on the package for including it.

As argued in this [issue](https://github.com/rust-lang/cargo/issues/16267), Cargo probably should do it by default.

**Comment by kornelski** [comment]
There's a trick that can be used to avoid having two archives: archive the necessary files first, and add the optional files last to the archive. Communicate where the necessary files end in the compressed archive (tar preserves order in which files were added, and gzip can stop decompressing at any point).

This way clients could use HTTP range requests to only download the prefix of the tarball and it would look like a valid tarball with only the subset of files. Clients unaware of this feature would simply download the whole archive.

The downside is that this would require adding the prefix size and a second checksum to the registry.
